core/src/main/java/org/kiwix/kiwixmobile/core/main/KiwixWebView.kt

```diff
@@ -0,0 +1,206 @@
+
+  @Inject
+  lateinit var zimReaderContainer: ZimReaderContainer
+
+  private val compositeDisposable = CompositeDisposable()
+
+  private fun setWindowVisibility(systemUiVisibility: Int) {
+    (context as Activity).window.decorView.systemUiVisibility = systemUiVisibility
+  }
+
+  init {
+    if (BuildConfig.DEBUG) {
+      WebView.setWebContentsDebuggingEnabled(true)
+    }
+    coreComponent.inject(this)
+    // Set the user agent to the current locale so it can be read with navigator.userAgent
+    settings.apply {
+      userAgentString = "${getCurrentLocale(context)}"
+      domStorageEnabled = true
+      javaScriptEnabled = true
+      loadWithOverviewMode = true
+      useWideViewPort = true
+      setInitialScale(INITIAL_SCALE)
+      builtInZoomControls = true
+      displayZoomControls = false
+      clearCache(true)
+      allowUniversalAccessFromFileURLs = true
+      setWebViewClient(webViewClient)
+    }
+    val client = KiwixWebChromeClient(callback, nonVideoView, videoView, this)
+    client.setOnToggledFullscreen { fullscreen ->
+      setWindowVisibility(
+        if (fullscreen) View.SYSTEM_UI_FLAG_LOW_PROFILE else View.SYSTEM_UI_FLAG_VISIBLE
+      )
+    }
+    webChromeClient = client
+  }
+
+  override fun performLongClick(): Boolean {
+    val result = hitTestResult
+    if (result.type == HitTestResult.SRC_ANCHOR_TYPE) {
+      callback.webViewLongClick(result.extra)
+      return true
+    }
+    return super.performLongClick()
+  }
+
+  override fun onCreateContextMenu(menu: ContextMenu) {
+    super.onCreateContextMenu(menu)
+    val result = hitTestResult
+    if (result.type == HitTestResult.IMAGE_ANCHOR_TYPE ||
+      result.type == HitTestResult.IMAGE_TYPE ||
+      result.type == HitTestResult.SRC_IMAGE_ANCHOR_TYPE
+    ) {
+      val saveMenu =
+        menu.add(0, 1, 0, resources.getString(R.string.save_media))
+      saveMenu.setOnMenuItemClickListener {
+        val msg = SaveHandler(zimReaderContainer).obtainMessage()
+        requestFocusNodeHref(msg)
+        true
+      }
+    }
+  }
+
+  override fun onAttachedToWindow() {
+    super.onAttachedToWindow()
+    compositeDisposable.add(
+      sharedPreferenceUtil.textZooms
+        .subscribe {
+          settings.textZoom = it
+        }
+    )
+  }
+
+  override fun onDetachedFromWindow() {
+    super.onDetachedFromWindow()
+    compositeDisposable.clear()
+  }
+
+  override fun onScrollChanged(l: Int, t: Int, oldl: Int, oldt: Int) {
+    super.onScrollChanged(l, t, oldl, oldt)
+    val windowHeight = if (measuredHeight > 0) measuredHeight else 1
+    val pages = contentHeight / windowHeight
+    val page = t / windowHeight
+    callback.webViewPageChanged(page, pages)
+  }
+
+  internal class SaveHandler(private val zimReaderContainer: ZimReaderContainer) :
+    Handler() {
+    private fun getDecodedFileName(url: String?, src: String?): String {
+      var fileName = ""
+      if (url != null) {
```

Comment:
we can kotlin this up, I think a chain of operators and using perhaps a [`takeIf`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/take-if.html) will streamline this

Reply:
basically how I think this could look is

```
return url?.SubstringAfterLast( etc
       .... newOperators
       .... onEachLine
```

Reply:
```
 private fun getDecodedFileName(url: String?, src: String?): String =
      url?.substringAfterLast("/").takeIf { it!!.contains(".") } ?: src?.substringAfterLast("/")
        ?.substringAfterLast("%3A") ?: ""
```
@macgills is this correct way?

Reply:
extremely close

first lets give it a quick format for readability

```
private fun getDecodedFileName(url: String?, src: String?): String =
  url?.substringAfterLast("/")
    .takeIf { it!!.contains(".") }
    ?: src?.substringAfterLast("/")
    ?.substringAfterLast("%3A") ?: ""
```

I think we should use the `missingDelimiterValue` to account for the cases where there is no `/` should that be a possible input. You can also remove the `!!` by using `?.takeIf`

Reply:
done